urllib3/connectionpool.py

```diff
@@ -568,25 +568,22 @@ def urlopen(self, method, url, body=None, headers=None, retries=None,
             # Close the connection. If a connection is reused on which there
             # was a Certificate error, the next request will certainly raise
             # another Certificate error.
-            if conn:
-                conn.close()
-                conn = None
+            conn = conn and conn.close()
+            release_conn = True
```

Comment:
> The BaseSSLError, CertificateError, SSLError exception handlers raise after setting release_conn to true, but that will bypass the finally block and the the conn (or None in this case) is never put back on to the queue. 

Why will that bypass the finally block?

> For the (TimeoutError, HTTPException, SocketError, ConnectionError) exception handler you don't raise (presumably because there may be retries) and you will go on to release the conn and make another call to urlopen (if retries). But you call it with release_conn=True (since you've just set it to true) - what happens if that call succeeds?

Yea this condition is a little hard to follow, we should try to add a test for it. Hmm.